src/main/java/org/eclipse/uprotocol/communication/InMemorySubscriber.java

```diff
@@ -160,17 +155,20 @@ public CompletionStage<UStatus> unsubscribe(UUri topic, UListener listener, Call
 
         final UnsubscribeRequest unsubscribeRequest = UnsubscribeRequest.newBuilder().setTopic(topic).build();
         
-        return RpcMapper.mapResponseToResult(
+        CompletionStage<RpcResult<UnsubscribeResponse>> result = RpcMapper.mapResponseToResult(
             // Send the unsubscribe request
-            rpcClient.invokeMethod(UNSUBSCRIBE_METHOD, UPayload.pack(unsubscribeRequest), options), 
-            UnsubscribeResponse.class)
-            
+            rpcClient.invokeMethod(UNSUBSCRIBE_METHOD, UPayload.pack(unsubscribeRequest), options),     
+                UnsubscribeResponse.class)
             // Then unregister the listener
-            .thenCompose(response -> {
+            .whenComplete((response, ex) -> mHandlers.remove(topic));
```

Comment:
We should remove it only when the unsubscribe request is successful.
```suggestion
            @Override
    public CompletionStage<UStatus> unsubscribe(UUri topic, UListener listener, CallOptions options) {
        Objects.requireNonNull(topic, "Unsubscribe topic missing");
        Objects.requireNonNull(listener, "listener missing");

        final UnsubscribeRequest unsubscribeRequest = UnsubscribeRequest.newBuilder().setTopic(topic).build();

        CompletionStage<RpcResult<UnsubscribeResponse>> result = RpcMapper.mapResponseToResult(
                // Send the unsubscribe request
                rpcClient.invokeMethod(UNSUBSCRIBE_METHOD, UPayload.pack(unsubscribeRequest), options),
                UnsubscribeResponse.class);

        // Create a CompletionStage that will handle the response from the unsubscribe request
        return transport.unregisterListener(topic, listener)
                .thenCombine(result, (status, response) -> {
                    if (response.isSuccess()) {
                        mHandlers.remove(topic); // Only remove the listener if unsubscribe is successful
                        return UStatus.newBuilder().setCode(UCode.OK).build();
                    }
                    return response.failureValue();
                });
    }
```

Reply:
As discussed with Steven, we can ignore UNSUBSCRIBE_PENDING to UNSUBSCRIBED state. Therefore, we don't need to keep the handler if the unsubscription fails. Hence, I am approving this PR.